# Predicting the Output Electrical Power of Combined Cycle Power Plant              with the help of 6 Years(2006-2011) data of the plant.
## Authour : Byri Manoj S

## Importing revelent packages

In [1]:
import tensorflow as tf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Importing Datasets

In [2]:
raw_data = pd.read_excel('Combined Cycle Power Plant Dataset.xlsx')
data = raw_data.copy()
data.head()

AT      V       AP     RH      PE
0  14.96  41.76  1024.07  73.17  463.26
1  25.18  62.96  1020.04  59.08  444.37
2   5.11  39.40  1012.16  92.14  488.56
3  20.86  57.32  1010.24  76.64  446.48
4  10.82  37.50  1009.23  96.62  473.90

## Data Preprocessing

In [3]:
data.rename(columns={'AT':'Temp', 'V':'Exhaust_Vaccum', 'AP':'Pressure', 'RH':'Humidity', 'PE':'Power_Output'})

Temp  Exhaust_Vaccum  Pressure  Humidity  Power_Output
0     14.96           41.76   1024.07     73.17        463.26
1     25.18           62.96   1020.04     59.08        444.37
2      5.11           39.40   1012.16     92.14        488.56
3     20.86           57.32   1010.24     76.64        446.48
4     10.82           37.50   1009.23     96.62        473.90
...     ...             ...       ...       ...           ...
9563  16.65           49.69   1014.01     91.00        460.03
9564  13.19           39.18   1023.67     66.78        469.62
9565  31.32           74.33   1012.92     36.48        429.57
9566  24.48           69.45   1013.86     62.39        435.74
9567  21.60           62.52   1017.23     67.87        453.28

[9568 rows x 5 columns]

In [4]:
data.shape

(9568, 5)

In [5]:
x = data.iloc[:,:-1].values
y = data.iloc[:, -1].values

In [6]:
from sklearn.model_selection import train_test_split as tts
x_train, x_test, y_train, y_test = tts(x, y, random_state = 143, test_size=0.2)

## Feature Scaling

In [7]:
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()
x_train = ss.fit_transform(x_train)
x_test = ss.transform(x_test)

## Buliding the Neural Network

In [8]:
ann_r = tf.keras.models.Sequential()

In [9]:
ann_r.add(tf.keras.layers.Dense(units=50, activation = 'relu'))
ann_r.add(tf.keras.layers.Dense(units=50, activation = 'relu'))
ann_r.add(tf.keras.layers.Dense(units=50, activation = 'relu'))
ann_r.add(tf.keras.layers.Dense(units=1))

In [10]:
ann_r.compile(optimizer='adam', loss='MeanSquaredError')

In [11]:
ann_r.fit(x_train, y_train, batch_size=32, epochs=100)

Epoch 1/100
240/240 [==============================] - 0s 369us/step - loss: 91697.4453
Epoch 2/100
240/240 [==============================] - 0s 354us/step - loss: 4812.5444
Epoch 3/100
240/240 [==============================] - 0s 357us/step - loss: 1527.8562
Epoch 4/100
240/240 [==============================] - 0s 357us/step - loss: 559.0094
Epoch 5/100
240/240 [==============================] - 0s 356us/step - loss: 246.1214
Epoch 6/100
240/240 [==============================] - 0s 351us/step - loss: 130.8932
Epoch 7/100
240/240 [==============================] - 0s 360us/step - loss: 80.2058
Epoch 8/100
240/240 [==============================] - 0s 356us/step - loss: 53.9722
Epoch 9/100
240/240 [==============================] - 0s 352us/step - loss: 40.7816
Epoch 10/100
240/240 [==============================] - 0s 355us/step - loss: 33.0637
Epoch 11/100
240/240 [==============================] - 0s 363us/step - loss: 28.1760
Epoch 12/100
240/240 [==============================]

240/240 [==============================] - 0s 351us/step - loss: 19.5321
Epoch 97/100
240/240 [==============================] - 0s 355us/step - loss: 19.3577
Epoch 98/100
240/240 [==============================] - 0s 352us/step - loss: 18.7872
Epoch 99/100
240/240 [==============================] - 0s 349us/step - loss: 19.1387
Epoch 100/100
240/240 [==============================] - 0s 353us/step - loss: 19.6712


In [12]:
ann_r.evaluate(x_train, y_train)

240/240 [==============================] - 0s 265us/step - loss: 18.3441


18.3441219329834

## Testing our network on unseen test dataset

In [13]:
y_pred = ann_r.predict(x_test)

### Comparing the predicted and actual outputs side-by-side

In [14]:
Evaluate = pd.DataFrame(data = y_pred, columns=['Predicted_Values'])
Evaluate['Targets'] = y_test

In [15]:
Evaluate

Predicted_Values  Targets
0           473.644318   470.85
1           477.024902   473.98
2           447.741058   452.21
3           469.122345   469.83
4           444.736023   444.89
...                ...      ...
1909        430.476471   432.98
1910        442.096191   440.78
1911        438.346893   439.77
1912        464.429321   462.21
1913        467.170105   464.62

[1914 rows x 2 columns]

In [16]:
from sklearn.metrics import *

In [17]:
r2_score(y_pred, y_test)

0.9358715037712251

# Our Artifical Brain is 94% accurate on the unseen dataset